src/librustc/ty/layout.rs

```diff
@@ -1478,13 +1475,37 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                         Ineligible(_) => false,
                     }
                 })
-                .map(|local| subst_field(info.field_tys[*local]));
+                .map(|local| subst_field(info.field_tys[*local]))
+                .map(|ty| self.layout_of(ty))
+                .map(|layout| {
+                    // Aggregate layouts can be partially initialized across
+                    // yields. The compiler currently warns on this instead of
+                    // giving a hard error (#60450).
+                    //
+                    // Since the compiler accepts this, generators are the only
+                    // place where a function can return while leaving a
+                    // partially initialized struct in memory (#60889).
+                    //
+                    // So, for example, even if a struct contains an uninhabited
+                    // field, we shouldn't mark the entire variant as
+                    // uninhabited. Individual aggregate fields should instead
+                    // be wrapped in MaybeUninit for this reason.
+                    //
+                    // Note that we only check this for variant only fields,
+                    // because promoted fields are unconditionally wrapped in
+                    // MaybeUninit above.
+                    let layout = layout?;
+                    if layout.is_aggregate() {
+                        self.layout_of(tcx.mk_maybe_uninit(layout.ty))
```

Comment:
IMO this is a bad idea: this won't properly trigger for newtypes of immediates (scalars/vectors).

Reply:
I guess the correct way would be with a type visitor then.

Or it's probably best to focus efforts on making #60450 a hard error, so we don't need this hack.

Reply:
Removing this in favor of #63230